introduce avatar traits system, migrate skeleton URL and avatar entities

birarda opened the pull request.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12018/HighFidelity-Beta-PR13803-ab8718a.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12018/HighFidelity-Beta-PR13803-ab8718a.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12018/HighFidelity-Beta-Interface-PR13803-ab8718a.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12018/HighFidelity-Beta-PR13803-ab8718a.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12018/HighFidelity-Beta-Interface-PR13803-ab8718a.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12049/HighFidelity-Beta-PR13803-2ed83ea.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12049/HighFidelity-Beta-PR13803-2ed83ea.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12049/HighFidelity-Beta-Interface-PR13803-2ed83ea.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12049/HighFidelity-Beta-PR13803-2ed83ea.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12049/HighFidelity-Beta-Interface-PR13803-2ed83ea.exe

birarda commented on interface/src/avatar/MyAvatar.cpp:
```diff
@@ -1287,7 +1287,6 @@ void MyAvatar::loadData() {
         // HACK: manually remove empty 'avatarEntityData' else legacy data may persist in settings file
         settings.remove("avatarEntityData");
     }
-    setAvatarEntityDataChanged(true);
```
This will be set now will each call to `updateAvatarEntity` above

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12061/HighFidelity-Beta-PR13803-72351f1.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12061/HighFidelity-Beta-PR13803-72351f1.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12061/HighFidelity-Beta-Interface-PR13803-72351f1.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12061/HighFidelity-Beta-PR13803-72351f1.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12061/HighFidelity-Beta-Interface-PR13803-72351f1.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12110/HighFidelity-Beta-PR13803-186df73.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12110/HighFidelity-Beta-PR13803-186df73.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12110/HighFidelity-Beta-Interface-PR13803-186df73.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12110/HighFidelity-Beta-PR13803-186df73.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12110/HighFidelity-Beta-Interface-PR13803-186df73.dmg

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12182/HighFidelity-Beta-PR13803-b5e2bda.apk

birarda commented:
### Test Plan

For these tests, you will need two machines running Interface from this PR build.

Because of a protocol version change, you will also need to connect to a domain running the version from this PR.

#### Skeleton URL

- From both connected Interface clients, start by confirming that you can see the other Avatar wearing the expected model.
- From both Interface clients change the avatar URL (manually via settings or using the marketplace) at least 5 times. Confirm that the expected change is seen by the other Interface client.
- From one of the Interface switch to another domain (which you will not be able to connect to because of a protocol version change) and back to the PR domain at least 5 times. Each time, confirm that each connected Interface client sees the other with the expected model.


#### Avatar URL Whitelist

- In the domain server settings, under Settings -> Avatars -> Advanced, set the "Avatars Allowed From:" to "mpassets.highfidelity.com"
- In the domain server settings, under Settings -> Avatars -> Advanced, set the "Replacement Avatar for disallowed avatars" to "http://mpassets.highfidelity.com/faf505f1-4fd1-4ed2-8909-816af246c48f-v1/M_VicGen.fst"
- Connect two Interface clients to the domain. Set the avatar URL of client A to "http://mpassets.highfidelity.com/96c747ab-f71b-44ee-8eb9-d19fc9593dda-v1/F_CatBur.fst" and set the avatar URL of client B to "http://example.com/dangerous-avatar.fst".
- Confirm that after connecting to the domain client A is wearing "http://mpassets.highfidelity.com/96c747ab-f71b-44ee-8eb9-d19fc9593dda-v1/F_CatBur.fst" and client B has had their avatar URL overridden and is wearing "http://mpassets.highfidelity.com/faf505f1-4fd1-4ed2-8909-816af246c48f-v1/M_VicGen.fst".


#### Avatar Entities

https://gist.githubusercontent.com/birarda/138892bffa2dda4e1e15d2bf70cc3c85/raw/0b51a2c56e3a86b4ea85c2641f1a3d3333f91955/test-trait-entities.js

- Use the linked script to test creation, update, and delete of avatar entities. First it creates a red box between the avatar's hands. 5 seconds later it updates the color of the box to blue. 5 seconds later it updates the color of the box to green. 5 seconds later it deletes the box. 5 seconds later it starts the cycle again.
- Using two Interface clients, connect a domain running the version from this PR. Run the test script as an Interface client script from each Interface (at different times). Allow the script to run for a couple of cycles each time and confirm you see the expected box created, updated, and deleted as described.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12182/HighFidelity-Beta-PR13803-b5e2bda.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12182/HighFidelity-Beta-Interface-PR13803-b5e2bda.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12182/HighFidelity-Beta-PR13803-b5e2bda.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12182/HighFidelity-Beta-Interface-PR13803-b5e2bda.exe

qlipdahlin commented:
Currently Testing

qlipdahlin commented:
@conklin94122 FAIL

Loaded interface and setup two local DS. Both interfaces saw each change, though one early instance of avatar not working with "http://mpassets.highfidelity.com/faf505f1-4fd1-4ed2-8909-816af246c48f-v1/M_VicGen.fst" and leaving pink mesh ball for avatar. Other avatars appeared to work so continued with test. Changing avatar test passed. Had guest change domains and return, with both interfaces being unable to see each other's avatar and only a pink mesh ball. Left it for 5~ minutes with no update and nothing in downloads or processing. Attempted to change avatars with no success. Guest left and came back and issue persisted.

Grabbed logs for both interfaces and owner server, noticed that the server log had a large group of about 5 kb log files, appearing as though there was a problem with the AC's. Checked task manager and only 6 AC's were present, but a new AC would appear about every 20-30 seconds and then disappear a few seconds after. The rest of the AC's appeared to work. Checked Nodes section and the Entity-Server-Script was left in queue.
[Logs_visitor.zip](https://github.com/highfidelity/hifi/files/2303025/Logs_visitor.zip)
[server_logs.zip](https://github.com/highfidelity/hifi/files/2303026/server_logs.zip)
[hifi-log_owner.txt](https://github.com/highfidelity/hifi/files/2303027/hifi-log_owner.txt)

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12201/HighFidelity-Beta-PR13803-284ae7f.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12201/HighFidelity-Beta-PR13803-284ae7f.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12201/HighFidelity-Beta-Interface-PR13803-284ae7f.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12201/HighFidelity-Beta-PR13803-284ae7f.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12201/HighFidelity-Beta-Interface-PR13803-284ae7f.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12209/HighFidelity-Beta-PR13803-7a0633d.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12209/HighFidelity-Beta-PR13803-7a0633d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12209/HighFidelity-Beta-Interface-PR13803-7a0633d.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12209/HighFidelity-Beta-PR13803-7a0633d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12209/HighFidelity-Beta-Interface-PR13803-7a0633d.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12216/HighFidelity-Beta-PR13803-2921597.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12216/HighFidelity-Beta-PR13803-2921597.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12216/HighFidelity-Beta-Interface-PR13803-2921597.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12216/HighFidelity-Beta-PR13803-2921597.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12216/HighFidelity-Beta-Interface-PR13803-2921597.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12222/HighFidelity-Beta-PR13803-e110511.apk

birarda commented on assignment-client/src/avatars/AvatarMixerSlave.cpp:
```diff
@@ -79,6 +79,111 @@ int AvatarMixerSlave::sendIdentityPacket(const AvatarMixerClientData* nodeData,
+        // there is definitely new traits data to send
+
+        // add the avatar ID to mark the beginning of traits for this avatar
+        bytesWritten += traitsPacketList.write(sendingNodeData->getNodeID().toRfc4122());
```
Can't here, because the client doesn't know about other avatar local IDs currently.

birarda commented on libraries/avatars/src/AvatarTraits.h:
```diff
@@ -0,0 +1,61 @@
+
+    using TraitInstanceID = QUuid;
+
+    inline bool isSimpleTrait(TraitType traitType) {
```
I considered this approach, but wanted to avoid people having to add traits at a specific place. I guess the `= FirstInstancedTrait` makes it clear enough

Atlante45 replied:
Good point looks fine for now, but when we add the rest, we should remember to add some spacing + comments so it's clearer.

birarda commented on libraries/avatars/src/AvatarData.cpp:
```diff
@@ -1846,26 +1820,109 @@ void AvatarData::processAvatarIdentity(const QByteArray& identityData, bool& ide
+    bytesWritten += destination.writePrimitive(traitType);
+
+    if (traitVersion > AvatarTraits::DEFAULT_TRAIT_VERSION) {
+        AvatarTraits::TraitVersion typedVersion = traitVersion;
```
I'll fix, it was a hold over from when versions were unsigned on the wire

birarda commented on assignment-client/src/avatars/AvatarMixerClientData.cpp:
```diff
@@ -87,6 +92,112 @@ int AvatarMixerClientData::parseData(ReceivedMessage& message) {
+                message.seek(message.getPosition() + traitSize);
+            }
+        } else {
+            AvatarTraits::TraitInstanceID instanceID = QUuid::fromRfc4122(message.readWithoutCopy(NUM_BYTES_RFC4122_UUID));
```
Definitely will want to change the avatar protocol to use local IDs, but have to work around that limitation with some parts of this change.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12230/HighFidelity-Beta-PR13803-c7b7355.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12230/HighFidelity-Beta-PR13803-c7b7355.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12230/HighFidelity-Beta-Interface-PR13803-c7b7355.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12230/HighFidelity-Beta-PR13803-c7b7355.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12230/HighFidelity-Beta-Interface-PR13803-c7b7355.exe

qlipdahlin commented:
Currently Testing

qlipdahlin commented:
@conklin94122 FAIL

**Skeleton URL**

- Clients connect and see correct avatar
- Changed avatars 5+ times on each client and avatar appeared as it should for each
- Avatar changed to one with wearables and was visible, changing to one without did not properly remove wearables and was visible from at least 1 client. Wearables app was empty.
- One client switched to their domain and back multiple times, each time returning as the correct avatar

**Avatar URL Whitelist**

- Changed "Avatars Allowed From:" to "mpassets.highfidelity.com"
- Changed "Replacement Avatar for disallowed avatars" to "http://mpassets.highfidelity.com/faf505f1-4fd1-4ed2-8909-816af246c48f-v1/M_VicGen.fst"
- Client A changed avatar to "http://mpassets.highfidelity.com/96c747ab-f71b-44ee-8eb9-d19fc9593dda-v1/F_CatBur.fst". Client B's "http://example.com/dangerous-avatar.fst" URL failed, only showing as woody avatar. Used another avatar URL that wasn't a mpassets "https://hifi-content.s3.amazonaws.com/jimi/avatar/Niconico/Jene_5.fst". Confirmed both avatars worked outside of the domain.
- Client A showed as Catburglar and Client B showed as VicGen.

**Avatar Entities**

- Used script on Avatar A, shows red box, blue box, green box, and removes above the head after 5 seconds for each and rotated multiple times with no apparent defect.
- Used script on Avatar B in same domain and confirmed both clients saw the proper creation, changes, and removal of the entities
- Client B closed interface and relaunched interface, noticed that entities had doubled up, 1 going through the proper cycle and others remaining as artifacts. Appears related to the wearable issue from earlier.


Appears to be a problem with entities sticking around when they should not. Based on test plan conditions, this should not occur and thus Fails this PR.

conklin94122 commented:
@birarda Ubuntu failed to build successfully as well

birarda commented:
Looking into the fail related to wearables. I have a feeling the failure in avatar entities is due to the script not cleaning up the avatar entity when it closes. I will add that so the behaviour is less confusing.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12242/HighFidelity-Beta-PR13803-3e2d4dc.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12242/HighFidelity-Beta-PR13803-3e2d4dc.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12242/HighFidelity-Beta-Interface-PR13803-3e2d4dc.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12242/HighFidelity-Beta-PR13803-3e2d4dc.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12242/HighFidelity-Beta-Interface-PR13803-3e2d4dc.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12247/HighFidelity-Beta-PR13803-8226ffb.apk

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12249/HighFidelity-Beta-PR13803-1723f3d.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12249/HighFidelity-Beta-PR13803-1723f3d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12249/HighFidelity-Beta-Interface-PR13803-1723f3d.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12249/HighFidelity-Beta-PR13803-1723f3d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12249/HighFidelity-Beta-Interface-PR13803-1723f3d.exe

qlipdahlin commented:
Currently Testing

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12255/HighFidelity-Beta-PR13803-7d8db48.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12255/HighFidelity-Beta-PR13803-7d8db48.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12255/HighFidelity-Beta-Interface-PR13803-7d8db48.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12255/HighFidelity-Beta-PR13803-7d8db48.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12255/HighFidelity-Beta-Interface-PR13803-7d8db48.exe

qlipdahlin commented:
@conklin94122 PASS

**Skeleton URL**

- Clients connect and see correct avatar
- Changed avatars 5+ times on each client and avatar appeared as it should for each
- Avatar changed to one with wearables and was visible, changing to one without properly removed the wearable. Tested multiple times and all wearables properly disappeared when avatar was changed.
- One client switched to their domain and back multiple times, each time returning as the correct avatar

**Avatar URL Whitelist**

- Changed "Avatars Allowed From:" to "mpassets.highfidelity.com"
- Changed "Replacement Avatar for disallowed avatars" to "http://mpassets.highfidelity.com/faf505f1-4fd1-4ed2-8909-816af246c48f-v1/M_VicGen.fst"
- Client A changed avatar to "http://mpassets.highfidelity.com/96c747ab-f71b-44ee-8eb9-d19fc9593dda-v1/F_CatBur.fst". Client B's "http://example.com/dangerous-avatar.fst" URL failed, only showing as woody avatar. Used another avatar URL that wasn't a mpassets "https://hifi-content.s3.amazonaws.com/jimi/avatar/Niconico/Jene_5.fst". Confirmed both avatars worked outside of the domain.
- Client A showed as Catburglar and Client B showed as VicGen.

**Avatar Entities**

- Used script on Avatar A, shows red box, blue box, green box, and removes above the head after 5 seconds for each and rotated multiple times with no apparent defect.
- Used script on Avatar B in same domain and confirmed both clients saw the proper creation, changes, and removal of the entities
- Removed script, closed interface, relaunched interface, and script did not continue. Script was modified to clean up when it was closed, but due to how closing the interface works, the script does not perform cleanup before client ends and thus would not cleanup afterwords.


All remaining issues dealt with and everything works as expected. PR passes.

divyaelango88 commented:
🎉
